contrib/src/main/scala/akka/stream/contrib/Valve.scala

```diff
@@ -0,0 +1,138 @@
+   */
+  def flip(mode: SwitchMode): Future[Boolean]
+}
+
+object Valve {
+  /**
+   * Factory for [[Valve]] instances.
+   *
+   */
+  def apply[A](): Valve[A] = Valve[A](SwitchMode.Open)
+
+  /**
+   * Java API: Factory for [[Valve]] instances.
+   *
+   */
+  def create[A](): Valve[A] = Valve[A](SwitchMode.Open)
+
+  /**
+   * Factory for [[Valve]] instances.
+   *
+   */
+  def apply[A](mode: SwitchMode): Valve[A] = new Valve[A](mode)
+
+  /**
+   * Java API: Factory for [[Valve]] instances.
+   *
+   */
+  def create[A](mode: SwitchMode): Valve[A] = Valve[A](mode)
+
+}
+
+/**
+ * Materializes into a [[ValveSwitch]] which provides a the method flip that stops or restarts the flow of elements passing through the stage. As long as the valve is closed it will backpressure.
+ *
+ * Note that closing the valve could result in one element being buffered inside the stage, and if the stream completes or fails while being closed, that element may be lost.
+ *
+ * @param mode state of the valve at the startup of the flow (by default Open)
+ */
+final class Valve[A](mode: SwitchMode) extends GraphStageWithMaterializedValue[FlowShape[A, A], ValveSwitch] {
+
+  val in: Inlet[A] = Inlet[A]("valve.in")
+
+  val out: Outlet[A] = Outlet[A]("valve.out")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ValveSwitch) = {
+    val logic = new ValveGraphStageLogic(shape, mode)
+    (logic, logic.switch)
+  }
+
+  private class ValveGraphStageLogic(shape: Shape, var mode: SwitchMode) extends GraphStageLogic(shape) with InHandler with OutHandler {
+    val switch = new ValveSwitch {
+
+      val callback = getAsyncCallback[(SwitchMode, Promise[Boolean])] {
+        case (flipToMode, promise) =>
+          val succeed = mode match {
+            case _ if flipToMode == mode => false
+
+            case Open =>
+              mode = SwitchMode.Close
+              true
+
+            case Close =>
+              bufferedElement.foreach(push(out, _))
+              bufferedElement = None
+
+              if (isAvailable(out)) {
+                // if a pull in has been requested we trigger pull(In)
+                pull(in)
+              }
+
+              mode = SwitchMode.Open
+              true
+          }
+
+          promise.success(succeed)
+      }
+
+      override def flip(flipToMode: SwitchMode): Future[Boolean] = {
+        val promise = Promise[Boolean]()
+        callback.invoke((flipToMode, promise))
+        promise.future
+      }
+    }
+    //Buffer the element to push() if the valve was close if a push was requested
+    var bufferedElement = Option.empty[A]
+
+    setHandlers(in, out, this)
+
+    override def onPush(): Unit = {
+      val element = grab(in)
+      if (isOpen) {
+        push(out, element)
+      } else {
+        bufferedElement = Some(element)
```

Comment:
no need for the buffer variable. Just don't grab() the element if you don't need it (and check for the presence of an element with `isAvailable`). See http://doc.akka.io/docs/akka/2.4/scala/stream/stream-customize.html#Port_states__InHandler_and_OutHandler for the state diagram for input ports.